ddtrace/tracer/stats.go

```diff
@@ -0,0 +1,331 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+//go:generate msgp -unexported -marshal=false -o=stats_msgp.go -tests=false
+
+package tracer
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/DataDog/sketches-go/ddsketch"
+)
+
+// aggregableSpan holds necessary information about a span that can be used to
+// aggregate statistics in a bucket.
+type aggregableSpan struct {
+	// key specifies the aggregation key under which this span can be placed into
+	// grouped inside a bucket.
+	key aggregation
+
+	Start, Duration int64
+	Error           int32
+	TopLevel        bool
+}
+
+// defaultStatsBucketSize specifies the default span of time that will be
+// covered in one stats bucket.
+var defaultStatsBucketSize = (10 * time.Second).Nanoseconds()
+
+// concentrator aggregates and stores statistics on incoming spans in time buckets,
+// flushing them occasionally to the underlying transport located in the given
+// tracer config.
+type concentrator struct {
+	// In specifies the channel to be used for feeding data to the concentrator.
+	// In order for In to have a consumer, the concentrator must be started using
+	// a call to Start.
+	In chan *aggregableSpan
+
+	// mu guards below fields
+	mu sync.Mutex
+
+	// buckets maintains a set of buckets, where the map key represents
+	// the starting point in time of that bucket, in nanoseconds.
+	buckets map[int64]*rawBucket
+
+	// stopped reports whether the concentrator is stopped (when non-zero)
+	stopped uint64
+
+	wg         sync.WaitGroup // waits for any active goroutines
+	bucketSize int64          // the size of a bucket in nanoseconds
+	stop       chan struct{}  // closing this channel triggers shutdown
+	cfg        *config        // tracer startup configuration
+}
+
+// newConcentrator creates a new concentrator using the given tracer
+// configuration c. It creates buckets of bucketSize nanoseconds duration.
+func newConcentrator(c *config, bucketSize int64) *concentrator {
+	return &concentrator{
+		In:         make(chan *aggregableSpan, 10000),
+		bucketSize: bucketSize,
+		stopped:    1,
+		buckets:    make(map[int64]*rawBucket),
+		cfg:        c,
+	}
+}
+
+// alignTs returns the provided timestamp truncated to the bucket size.
+// It gives us the start time of the time bucket in which such timestamp falls.
+func alignTs(ts, bucketSize int64) int64 { return ts - ts%bucketSize }
+
+// Start starts the concentrator. A started concentrator needs to be stopped
+// in order to gracefully shut down, using Stop.
+func (c *concentrator) Start() {
+	if atomic.SwapUint64(&c.stopped, 0) == 0 {
+		// already running
```

Comment:
nit: Add a log here? I guess you're not supposed to start the concentrator twice.

Reply:
Sure, why not.